TimeOfFlight/src/TOFUtils.cc

```diff
@@ -2,145 +2,363 @@
+}
+
+
+double TOFUtils::getTPCOuterR(){
+    auto& detector = Detector::getInstance();
+    DetElement tpcDet = detector.detector("TPC");
+    FixedPadSizeTPCData* tpc = tpcDet.extension <FixedPadSizeTPCData>();
+    return tpc->rMaxReadout/dd4hep::mm;
+}
+
+
+EVENT::TrackerHit* TOFUtils::getSETHit(EVENT::Track* track, double tpcOuterR){
+    vector<TrackerHit*> hits = track->getTrackerHits();
+    TrackerHit* lastHit = hits.back();
+    Vector3D pos (lastHit->getPosition());
+
+    if ( pos.rho() > tpcOuterR ) return lastHit;
+    return nullptr;
+}
+
+
+std::vector<EVENT::CalorimeterHit*> TOFUtils::selectFrankEcalHits( EVENT::Cluster* cluster, EVENT::Track* track, int maxEcalLayer, double bField ){
+    vector<CalorimeterHit*> selectedHits;
+
+    const TrackState* tsEcal = track->getTrackState(TrackState::AtCalorimeter);
+    Vector3D trackPosAtEcal ( tsEcal->getReferencePoint() );
+    Vector3D trackMomAtEcal = TOFUtils::getHelixMomAtTrackState(*tsEcal, bField);
+
+    for (int l=0; l<maxEcalLayer; ++l){
+        // find the closest hit to the linearly extrapolated track inside ecal for each layer
+        // OPTIMIZE: this is probably not the most efficient way to loop over ALL hits "maxEcalLayer" times
+        CalorimeterHit* selectedHit = nullptr;
+        double closestDistanceToLine = numeric_limits<double>::max();
+        for ( auto hit : cluster->getCalorimeterHits() ){
+            CHT hitType( hit->getType() );
+            bool isECALHit = ( hitType.caloID() == CHT::ecal );
+            if ( (!isECALHit) || ( int( hitType.layer() ) != l) ) continue;
+
+            Vector3D hitPos( hit->getPosition() );
+            double dToLine = (hitPos - trackPosAtEcal).cross(trackMomAtEcal.unit()).r();
+            if (dToLine < closestDistanceToLine){
+                closestDistanceToLine = dToLine;
+                selectedHit = hit;
+            }
+        }
+        if ( selectedHit != nullptr ) selectedHits.push_back(selectedHit);
+    }
+    return selectedHits;
+}
+
+
+std::vector<EVENT::Track*> TOFUtils::getSubTracks(EVENT::Track* track){
+    vector<Track*> subTracks;
+    subTracks.push_back(track);
+
+    int nSubTracks = track->getTracks().size();
+    if (nSubTracks <= 1) return subTracks;
+
+    int nTPCHits = track->getSubdetectorHitNumbers()[(ILDDetID::TPC)*2-1];
+    int nSubTrack0Hits = track->getTracks()[0]->getTrackerHits().size();
+    int nSubTrack1Hits = track->getTracks()[1]->getTrackerHits().size();
+
+    //OPTIMIZE: this is not reliable, but seems no other way is possible.
+    //Read documentation in the header file for details.
+    int startIdx;
+    if( std::abs(nTPCHits - nSubTrack0Hits) <= 1  ) startIdx = 1;
+    else if ( std::abs(nTPCHits - nSubTrack1Hits) <= 1 ) startIdx = 2;
+    else{
+        // this shouldn't happen in princinple at all...
+        streamlog_out(WARNING)<<"Can't understand which subTrack is responsible for the first TPC hits! Skip adding subTracks."<<std::endl;
```

Comment:
I comment here on the whole `getSubTracks()` thing...

_"I assume this is due to the presence of curlers and the fit not being able to handle that(?)."_

Yes.

The fitter needs **sorted** hits. 
In the current production for this we sort tracker hits by the radius,
which means we can't fit more than half a turn at the time,
because after half a turn the radius of the tracker hit may have a lower radius and the whole "sorting" doesn't work...

So, the way it is done is that, we have many fitted subTracks (half-turns) which are later combined in a single track if some merging condition is fulfilled.

The final single `Track` contains only hits for the first half-turn and **does not** contain all the hits from all subTracks.
But it stores the subTracks itself, so I need to access the subTracks to get all the hits, not only for the first half-turn.

I have 1st half-turn hits from the "The final single `Track`".

So to get all the hits correctly I need to get the index of the 2nd subTrack in the `track->getTracks()` vector.

And this is not trivial, because sometimes subTracks are written as:
subTracks = {"VXD+SIT subTrack", "1st subTrack", "2nd subTrack", "3rd subTrack", ...}
and sometimes as:
subTracks = {&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; "1st subTrack", "2nd subTrack", "3rd subTrack", ...}
So I need to guess...

I am guessing by comparing the number of hits inside the "`1st subTrack`" and "The final single `Track`" which in principle should agree. But they sometimes diverge for +-1 hit...
Could it happen that they diverge for more than 1 hit? I have no idea...

So in this case I just return the hits of the "The final single `Track`" with a warning...
In principle this is something I can also investigate... But some times is needed..

I am not sure what do you mean by "proper" error in this case...
Should we ignore the track at all in such a case and just return e.g. `{0., 0., 0.}`?
Or crash a program?
Or...

I will change the `WARNING` to `ERROR` in the log output for the time being...

Reply:
P.S. in principle for the fit, we can think of a smarter way to sort tracker hits...
E.g. sorting them along the "initial helix" before the fit would solve this problem and be much better logic-wise.
But then we need to think how to define startHit, middleHit and endHit to initialize the helix before the fit itself...

We can't simply sort by `z`, because for the very transverse tracks, fluctuations along `z` can mess things up, but maybe we can use some smart combination of the `z` and `R` instead...
These are some raw thoughts for further improvements... 

But this is something too far away from the TOFEstimator scope...

Reply:
@tmadlener I also checked if the " _Could it happen that they diverge for more than 1 hit? I have no idea..._" case happens at all.

And it does in 0.01% of pfos:

![subTracks](https://user-images.githubusercontent.com/28848011/136233523-7819187a-9817-4a88-87c6-da8f1814eef7.png)

Again, I have no good understanding why... One need to very carefully check the code, where the individual subTracks are created and filled with hits and compare it to the main Track case... And it sounds painful and very time consuming..

In the mean time, very small fraction of events is affected...
And even if we understand why this is happening, there is no clear sight that we could fix it within TOFEstimators...

Any thoughts?